src/tensor_ops/sigmoid.rs

```diff
@@ -21,20 +15,20 @@ pub struct SigmoidKernelOp;
 /// let t = dev.tensor([-1.0, 0.0, 1.0, 2.0]);
 /// let r = t.sigmoid();
 /// ```
-pub fn sigmoid<S: Shape, E: Dtype, D: UnaryKernel<SigmoidKernelOp, E>, T: Tape<E, D>>(
+pub fn sigmoid<S: Shape, E: Dtype, D: Device<E>, T: Tape<E, D>>(
     t: Tensor<S, E, D, T>,
 ) -> Tensor<S, E, D, T> {
     t.sigmoid()
 }
 
-impl<S: Shape, E: Dtype, D: UnaryKernel<SigmoidKernelOp, E>, T: Tape<E, D>> Tensor<S, E, D, T> {
+impl<S: Shape, E: Dtype, D: Device<E>, T: Tape<E, D>> Tensor<S, E, D, T> {
     /// See [sigmoid]
     pub fn sigmoid(self) -> Self {
         self.try_sigmoid().unwrap()
     }
     /// See [sigmoid]
     pub fn try_sigmoid(self) -> Result<Self, D::Err> {
-        try_unary_op(SigmoidKernelOp, self)
+        self.try_negate()?.try_exp()?.try_add(E::ONE)?.try_recip()
```

Comment:
Is calling 4 kernels for an activation function a good idea? This is bound to be a little slower, and will use 4 times more memory during training,

Reply:
Without operator fusion probably not a good idea - I think it could actually be much much slower since it'll have to alloc 4 times instead of 1 times.